memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -208,44 +196,70 @@ func (r *ReconcileMemcached) deploymentForMemcached(m *cachev1alpha1.Memcached)
 				},
 				Spec: corev1.PodSpec{
 					Containers: []corev1.Container{{
-						Image:   "memcached:1.4.36-alpine",
-						Name:    "memcached",
-						Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
-						Ports: []corev1.ContainerPort{{
-							ContainerPort: 11211,
-							Name:          "memcached",
-						}},
+						Name:    name,
+						Image:   image,
+						Ports:   ports,
+						Command: command,
 					}},
 				},
 			},
-		},
+		}
+
+		return
+	}
+
+	if *deployment.Spec.Replicas != replicas {
+		reqLogger.Info("Fixing replica count in Deployment", "From", deployment.Spec.Replicas, "To", replicas)
+		deployment.Spec.Replicas = &replicas
+	}
+	if !reflect.DeepEqual(deployment.Spec.Template.ObjectMeta.Labels, ls) {
+		reqLogger.Info("Fixing labels in Deployment Pod template", "From", deployment.Spec.Template.ObjectMeta.Labels, "To", ls)
+		deployment.Spec.Template.ObjectMeta.Labels = ls
+	}
+	if deployment.Spec.Template.Spec.Containers[0].Name != name {
+		reqLogger.Info("Fixing container name in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Name, "To", name)
+		deployment.Spec.Template.Spec.Containers[0].Name = name
+	}
+	if deployment.Spec.Template.Spec.Containers[0].Image != image {
+		reqLogger.Info("Fixing container image in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Image, "To", image)
+		deployment.Spec.Template.Spec.Containers[0].Image = image
+	}
+	if !reflect.DeepEqual(deployment.Spec.Template.Spec.Containers[0].Ports, ports) {
+		reqLogger.Info("Fixing container ports in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Ports, "To", ports)
+		deployment.Spec.Template.Spec.Containers[0].Ports = ports
 	}
-	// Set Memcached instance as the owner of the Deployment.
-	controllerutil.SetControllerReference(m, dep, r.scheme)
```

Comment:
IMHO: Is not a good approach not set the ownership when the object is created, What happens if has some issue be faced and it cannot be updated after for example?

Reply:
@camilamacedo86 I think @grdryn is right. If you look at the implementation of [`SetControllerReference()`](https://github.com/kubernetes-sigs/controller-runtime/blob/16c93b06b0a8cb99e002529cb1b43dce5e489baf/pkg/controller/controllerutil/controllerutil.go#L56) it is just working against the object in memory, doesn't make any API calls, and is idempotent since it [handles the case](https://github.com/kubernetes-sigs/controller-runtime/blob/16c93b06b0a8cb99e002529cb1b43dce5e489baf/pkg/controller/controllerutil/controllerutil.go#L73) of trying to set the same controller owner reference multiple times.

I think the correct approach is to leave it as is and always try to set it. The current approach would have the following properties:
1. It would set the correct controller owner reference on creation
2. It would keep the correct controller owner reference on update if the existing controller owner reference was unchanged in the previous update.
3. It would set the correct controller owner reference on update if the controller owner reference was deleted from the previous update.
4. If would error if the controller owner reference is different than the one we're trying to set.

Reply:
Hi @grdryn,

Sorry about that. when I open the file without the git diff was clear for me that it will **always** set the owner. So, I am closing this one. 

c/c @joelanford